src/rust/en.mangairo/src/parser.rs

```diff
@@ -0,0 +1,354 @@
+use aidoku::{
+	error::Result, helpers::uri::QueryParameters, prelude::*, std::html::Node, std::String,
+	std::Vec, Chapter, Filter, FilterType, Manga, MangaContentRating, MangaStatus, MangaViewer,
+	Page,
+};
+extern crate alloc;
+use alloc::string::ToString;
+
+pub const BASE_URL: &str = "https://w.mangairo.com";
+pub const USER_AGENT: &str = "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_3_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36";
+
+pub fn parse_manga_list(html: Node, page: i32) -> (Vec<Manga>, bool) {
+	let mut result: Vec<Manga> = Vec::new();
+	for page in html.select(".story-item").array() {
+		let obj = page.as_node().expect("node array");
+
+		let id = obj.select(".story-name a").attr("href").read();
+		let title = obj.select(".story-name a ").text().read();
+		let cover = obj.select(".story-list-img img").attr("src").read();
+
+		if !id.is_empty() && !title.is_empty() && !cover.is_empty() {
+			result.push(Manga {
+				id,
+				cover,
+				title,
+				..Default::default()
+			});
+		}
+	}
+
+	// Example: 'Total: 38,202 stories'
+	let mut total_str = html.select(".quantitychapter").text().read();
+	total_str = total_str.replace("Total: ", "");
+	total_str = total_str.replace(" stories", "");
+	total_str = total_str.chars().filter(|&c| c != ',').collect();
+
+	let has_more = total_str
+		.parse::<i32>()
+		.map_or(false, |value| value > result.len() as i32 * page);
+	(result, has_more)
+}
+
+pub fn parse_manga_details(html: Node, id: String) -> Result<Manga> {
+	let title = html
+		.select(".breadcrumbs p span a span")
+		.last()
+		.text()
+		.read();
+	let cover = html.select(".avatar").attr("src").read();
+	let description = html
+		.select("div#story_discription p")
+		.text()
+		.read()
+		.trim()
+		.to_string();
+	let status_str = html
+		.select(".story_info_right li:nth-child(5) a")
+		.text()
+		.read()
+		.to_lowercase();
+
+	let url = format!("{}", &id);
```

Comment:
ids should **not** be full urls. the best way to handle this is to string the url out so you just have the `/relative/path`, and then you can prepend the `BASE_URL` to get the full url. This makes it easier to fix if the site ever changes domains.

Reply:
I ran into a problem with that because the latest manga use URL `w.mangairo.com`, but the older manga use `chap.mangairo.com`. I don't remember exactly what was the problem I ran into, I'll revisit this again.

Reply:
is it possible to only use `chap.mangairo.com` then?

Reply:
Should be... I'll need to do some refactoring though.

Reply:
Unfortunately, it doesn't work. `https://chap.mangairo.com/story-de293390/chapter-1` will redirect to `https://w.mangairo.com/story-de293390`, breaking method `get_page_list`

Reply:
alright then it's probably fine to just leave it with the full urls